cpp/include/raft/spatial/knn/detail/ivf_flat/warpsort_topk.cuh

```diff
@@ -0,0 +1,862 @@
+    namely [start, end) of in. The idx is inferred from start.
+    This function should be called only once to add all items, and should not be
+    used together with the second form of add().
+    The second one is add(T val, IdxT idx), and it adds only one item pair.
+    Note that the range [start, end) is for the whole block of threads, that is,
+    each thread in the same block should get the same start/end.
+    In contrast, the parameters of the second form are for only one thread,
+    so each thread must get different val/idx.
+
+    After adding is finished, function done() should be called. And finally,
+    dump() is used to get the top-k result.
+
+    Example:
+      __global__ void kernel() {
+        WarpSortBlockWide<WarpBitonic, ...> queue(...);
+
+        // way 1, [0, len) is same for the whole block
+        queue.add(in, 0, len);
+        // way 2, each thread gets its own val/idx pair
+        for (IdxT i = threadIdx.x; i < len, i += blockDim.x) {
+          queue.add(in[i], idx[i]);
+        }
+
+        queue.done();
+        queue.dump(out, out_idx);
+     }
+
+     int smem_size = calc_smem_size_for_block_wide<T>(...);
+     kernel<<grid_dim, block_dim, smem_size>>>();
+
+
+  3. class WarpSelect and class WarpBitonic
+    These two classes can be regarded as fixed sized priority queue for a warp.
+    Usage is similar to class WarpSortBlockWide.
+    Two types of add() functions are provided, and also note that [start, end) is
+    for a whole warp, while val/idx is for a thread.
+    No shared memory is needed.
+
+    Example:
+      __global__ void kernel() {
+        WarpBitonic<...> queue(...);
+        int warp_id = threadIdx.x / WarpSize;
+        int lane_id = threadIdx.x % WarpSize;
+
+        // way 1, [0, len) is same for the whole warp
+        queue.add(in, 0, len);
+        // way 2, each thread gets its own val/idx pair
+        for (IdxT i = lane_id; i < len, i += WarpSize) {
+          queue.add(in[i], idx[i]);
+        }
+
+        queue.done();
+        // each warp outputs to a different offset
+        queue.dump(out+ warp_id * k * sizeof(T), out_idx+ warp_id * k * sizeof(IdxT));
+      }
+ */
+
+namespace raft::spatial::knn::detail::ivf_flat {
+
+static constexpr int kMaxCapacity = 512;
+
+namespace {
+
+template <bool greater, typename T>
+__device__ inline bool is_greater_than(T val, T baseline)
+{
+  if constexpr (greater) { return val > baseline; }
+  if constexpr (!greater) { return val < baseline; }
+}
+
+int calc_capacity(int k)
+{
+  int capacity = isPo2(k) ? k : (1 << (log2(k) + 1));
+  if (capacity < WarpSize) { capacity = WarpSize; }
+  return capacity;
+}
+
+}  // namespace
+
+template <int capacity, bool greater, typename T, typename IdxT>
+class WarpSort {
+ public:
+  __device__ WarpSort(IdxT k, T dummy) : lane_(threadIdx.x % WarpSize), k_(k), dummy_(dummy)
+  {
+    static_assert(capacity >= WarpSize && isPo2(capacity));
```

Comment:
Is there any relationship between `capacity` and `k` that we should check?

Reply:
Yes, it's ensured in the calling code that `k <= capacity`, but we cannot use raft's `ASSERT` in the device code to check the value of `k` at runtime.